[View in Colaboratory](https://colab.research.google.com/github/RomanShekhovtsov/ml-intro/blob/master/7_final.ipynb)

In [0]:
from time import time

def t0():
    global t_0
    t_0 = time()

def ptime():
    print( "TIME: ", round( time() - t_0, 3) )
  

In [2]:
import pandas as pd
import numpy as np

# 1. Считайте таблицу с признаками из файла features.csv с помощью кода, приведенного выше
gdrive_file_id = '13oDRm2q6vsh6OtHJB56tdY82-DMSpz7s'
features_url = 'https://drive.google.com/uc?export=download&confirm=no_antivirus&id=' + gdrive_file_id

d = pd.read_csv( features_url, index_col='match_id', compression='zip' )
#features.head()

# Удалите признаки, связанные с итогами матча (они помечены в описании данных как отсутствующие в тестовой выборке).
features = d.drop(d.columns[-6:], 1)
features.head()

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
0         1430198770           7       11         5   2098     1489     20   
1         1430220345           0       42         4   1188     1033      9   
2         1430227081           7       33         4   1319     1270     22   
3         1430263531           1       29         4   1779     1056     14   
4         1430282290           7       13         4   1431     1090      8   

          r1_kills  r1_deaths  r1_items          ...           \
match_id                                         ...            
0                0          0         7          ...            
1                0          1        12          ...            
2                0          0        12          ...            
3                0          0         5          ...            
4                1          0         8          ...            

          radiant_ward_sentry_count  radiant_first_ward_time  \
match_id                                                       
0                                 0                     35.0   
1                                 0                    -20.0   
2                                 1                    -39.0   
3                                 0                    -30.0   
4                                 0                     46.0   

          dire_bottle_time  dire_courier_time  dire_flying_courier_time  \
match_id                                                                  
0                    103.0              -84.0                     221.0   
1                    149.0              -84.0                     195.0   
2                     45.0              -77.0                     221.0   
3                    124.0              -80.0                     184.0   
4                    182.0              -80.0                     225.0   

          dire_tpscroll_count  dire_boots_count  dire_ward_observer_count  \
match_id                                                                    
0                           3                 4                         2   
1                           5                 4                         3   
2                           3                 4                         3   
3                           0                 4                         2   
4                           6                 3                         3   

          dire_ward_sentry_count  dire_first_ward_time  
match_id                                                
0                              2                 -52.0  
1                              1                  -5.0  
2                              1                  13.0  
3                              0                  27.0  
4                              0                 -16.0  

[5 rows x 102 columns]

In [3]:
features.describe()

start_time    lobby_type       r1_hero      r1_level         r1_xp  \
count  9.723000e+04  97230.000000  97230.000000  97230.000000  97230.000000   
mean   1.444232e+09      2.630999     51.517104      3.442672   1233.405801   
std    5.515393e+06      2.835761     32.564211      1.111741    566.588895   
min    1.430199e+09      0.000000      1.000000      0.000000      0.000000   
25%    1.440815e+09      1.000000     22.000000      3.000000    767.000000   
50%    1.446338e+09      1.000000     50.000000      3.000000   1175.000000   
75%    1.448829e+09      7.000000     75.000000      4.000000   1704.000000   
max    1.450313e+09      7.000000    112.000000      6.000000   3319.000000   

            r1_gold         r1_lh      r1_kills     r1_deaths      r1_items  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean    1147.899702     11.231996      0.357009      0.362285      8.271315   
std      464.111662      9.041620      0.663889      0.626704      2.497575   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%      746.000000      2.000000      0.000000      0.000000      7.000000   
50%     1113.000000     11.000000      0.000000      0.000000      8.000000   
75%     1479.000000     19.000000      1.000000      1.000000     10.000000   
max     4332.000000     47.000000      8.000000      5.000000     34.000000   

               ...           radiant_ward_sentry_count  \
count          ...                        97230.000000   
mean           ...                            0.716250   
std            ...                            0.725331   
min            ...                            0.000000   
25%            ...                            0.000000   
50%            ...                            1.000000   
75%            ...                            1.000000   
max            ...                           25.000000   

       radiant_first_ward_time  dire_bottle_time  dire_courier_time  \
count             95394.000000      81087.000000       96554.000000   
mean                 -6.875747        127.215028         -80.191893   
std                  39.508650         62.442018          15.261950   
min                -236.000000        -45.000000         -90.000000   
25%                 -31.000000         83.000000         -86.000000   
50%                 -15.000000        131.000000         -84.000000   
75%                   9.000000        165.000000         -79.000000   
max                 300.000000        300.000000         296.000000   

       dire_flying_courier_time  dire_tpscroll_count  dire_boots_count  \
count              71132.000000         97230.000000      97230.000000   
mean                 214.870536             2.965566          3.349553   
std                   34.137158             1.907288          1.155609   
min                  180.000000             0.000000          0.000000   
25%                  185.000000             2.000000          3.000000   
50%                  203.000000             3.000000          3.000000   
75%                  238.000000             4.000000          4.000000   
max                  300.000000            21.000000          9.000000   

       dire_ward_observer_count  dire_ward_sentry_count  dire_first_ward_time  
count              97230.000000            97230.000000          95404.000000  
mean                   2.448339                0.689119             -6.901922  
std                    0.813459                0.710122             40.701397  
min                    0.000000                0.000000            -84.000000  
25%                    2.000000                0.000000            -31.000000  
50%                    2.000000                1.000000            -16.000000  
75%                    3.000000                1.000000              8.000000  
max                    9.000000               13.000000            300.000000  

[8 rows x 102 columns]

In [4]:
# 2. Проверьте выборку на наличие пропусков с помощью функции count(), которая для каждого столбца показывает число заполненных значений.
# Много ли пропусков в данных?
def features_with_miss( features ):
    c = features.count()
    print( 'max:', max(c) )
    print( 'misses:', c[c<max(c)].sort_values(), sep='\n')
  
features_with_miss( features )

max: 97230
misses:
first_blood_player2            53243
radiant_flying_courier_time    69751
dire_flying_courier_time       71132
first_blood_time               77677
first_blood_team               77677
first_blood_player1            77677
dire_bottle_time               81087
radiant_bottle_time            81539
radiant_first_ward_time        95394
dire_first_ward_time           95404
radiant_courier_time           96538
dire_courier_time              96554
dtype: int64


##Запишите названия признаков, имеющих пропуски, и попробуйте для любых двух из них дать обоснование, почему их значения могут быть пропущены.

###Признаки, имеющие пропуски
```
first_blood_player2            53243
radiant_flying_courier_time    69751
dire_flying_courier_time       71132
first_blood_time               77677
first_blood_team               77677
first_blood_player1            77677
dire_bottle_time               81087
radiant_bottle_time            81539
radiant_first_ward_time        95394
dire_first_ward_time           95404
radiant_courier_time           96538
dire_courier_time              96554
```
###Обоснование

Очевидно, в первые 5 мин матча не происходили события, описываемые данными признаками.

При этом по событию **first_blood** мы видим аномалию признака **first_blood_player2** - заполнено 53243 строки.
У остальных признаков **first_blood**, например, **first_blood_time**, заполнено 77677 строк.
Возможная причина (https://otvet.mail.ru/question/83833036):

First blood” может быть засчитан если игрок был убит башней или крипами принадлежащими к команде соперника при условии что это произошло в короткий промежуток времени после нанесения урона игроком противоположной команды


In [5]:
# 3. Замените пропуски на нули с помощью функции fillna(). 
features = features.fillna(0)
features_with_miss( features )

# На самом деле этот способ является предпочтительным для логистической регрессии, поскольку он позволит пропущенному значению не вносить никакого вклада в предсказание.
# Для деревьев часто лучшим вариантом оказывается замена пропуска на очень большое или очень маленькое значение — в этом случае при построении разбиения вершины
# можно будет отправить объекты с пропусками в отдельную ветвь дерева. Также есть и другие подходы — например, замена пропуска на среднее значение признака.
# Мы не требуем этого в задании, но при желании попробуйте разные подходы к обработке пропусков и сравните их между собой.

max: 97230
misses:
Series([], dtype: int64)


In [6]:
# 4. Какой столбец содержит целевую переменную? Запишите его название.
target = d.radiant_win
target.head()

match_id
0    1
1    1
2    0
3    0
4    0
Name: radiant_win, dtype: int64

In [0]:
randomState = 1 #заменить на Nonе для проверки с вариациями
nJobs = 1 #сколько ядер используем

# 5. Забудем, что в выборке есть категориальные признаки, и попробуем обучить градиентный бустинг над деревьями на имеющейся матрице "объекты-признаки". 
# Зафиксируйте генератор разбиений для кросс-валидации по 5 блокам (KFold), не забудьте перемешать при этом выборку (shuffle=True),
# поскольку данные в таблице отсортированы по времени, и без перемешивания можно столкнуться с нежелательными эффектами при оценивании качества. 
from sklearn.model_selection import KFold, cross_val_score
kf = KFold( 5, shuffle=True, random_state=randomState )

# Оцените качество градиентного бустинга (GradientBoostingClassifier) с помощью данной кросс-валидации,
# попробуйте при этом разное количество деревьев (как минимум протестируйте следующие значения для количества деревьев: 10, 20, 30).
from sklearn.ensemble import GradientBoostingClassifier
for n in (10, 20, 30):
    t0()
    gbc = GradientBoostingClassifier( n_estimators=n, random_state=randomState )
    scores = cross_val_score( gbc, features, target, cv=kf, scoring='roc_auc', n_jobs=nJobs )
    print( np.mean( scores ) )
    ptime()

###Долго ли настраивались классификаторы?###
Результат на Intel Core i5 (4 core, n_jobs=4):
```
10 0.6648329228049101 TIME:  21.174
20 0.6821140369500347 TIME:  38.3
30 0.6896947542059905 TIME:  54.542
```
###Достигнут ли оптимум на испытанных значениях параметра n_estimators, или же качество, скорее всего, продолжит расти при дальнейшем его увеличении?###
Качество, скорее всего, продолжит расти при дальнейшем его увеличении.

In [0]:
# Подход 2: логистическая регрессия

# Важно: не забывайте, что линейные алгоритмы чувствительны к масштабу признаков!
# Может пригодиться sklearn.preprocessing.StandartScaler.
from sklearn.preprocessing import StandartScaler
scaler = StandardScaler()
scaled_features = scaler.fit_transform( features )

# Оцените качество логистической регрессии (sklearn.linear_model.LogisticRegression с L2-регуляризацией) 
# с помощью кросс-валидации по той же схеме, которая использовалась для градиентного бустинга. 

# Подберите при этом лучший параметр регуляризации (C). 

# Какое наилучшее качество у вас получилось?
# Как оно соотносится с качеством градиентного бустинга? 
# Чем вы можете объяснить эту разницу?
# Быстрее ли работает логистическая регрессия по сравнению с градиентным бустингом?

# Среди признаков в выборке есть категориальные, которые мы использовали 
# как числовые, что вряд ли является хорошей идеей.
# Категориальных признаков в этой задаче одиннадцать:
# lobby_type и r1_hero, r2_hero, ..., r5_hero, d1_hero, d2_hero, ..., d5_hero. 
# Уберите их из выборки, и проведите кросс-валидацию для логистической регрессии
# на новой выборке с подбором лучшего параметра регуляризации. 
# Изменилось ли качество? Чем вы можете это объяснить?

# Выясните из данных, сколько различных идентификаторов героев существует 
# в данной игре (вам может пригодиться фукнция unique или value_counts).

# Воспользуемся подходом "мешок слов" для кодирования информации о героях. 
# Пусть всего в игре имеет N различных героев.
# Сформируем N признаков, при этом i-й будет равен нулю, если i-й герой
# не участвовал в матче; единице, если i-й герой играл за команду Radiant;
# минус единице, если i-й герой играл за команду Dire.

# Добавьте полученные признаки к числовым, которые вы использовали во втором пункте данного этапа.

# Проведите кросс-валидацию для логистической регрессии на новой выборке с подбором лучшего параметра регуляризации.

# Какое получилось качество? Улучшилось ли оно? Чем вы можете это объяснить?

# Постройте предсказания вероятностей победы команды Radiant
# для тестовой выборки с помощью лучшей из изученных моделей
# (лучшей с точки зрения AUC-ROC на кросс-валидации).

# Убедитесь, что предсказанные вероятности адекватные — находятся на отрезке
# [0, 1], не совпадают между собой (т.е. что модель не получилась константной).